File: Android/app/src/main/java/io/nekohasekai/sfa/repository/SubscriptionMetadata.kt
package io.nekohasekai.sfa.repository

import io.nekohasekai.sfa.database.TypedProfile
import java.net.URI
import kotlin.math.max

data class SubscriptionMetadata(
    val upload: Long? = null,
    val download: Long? = null,
    val total: Long? = null,
    val expireAtMillis: Long? = null,
    val updateIntervalMinutes: Int? = null,
    val webPageUrl: String? = null,
) {
    fun isEmpty(): Boolean = upload == null &&
        download == null &&
        total == null &&
        expireAtMillis == null &&
        updateIntervalMinutes == null &&
        webPageUrl == null

    fun replaceOn(profile: TypedProfile) {
        profile.subscriptionUpload = upload ?: -1L
        profile.subscriptionDownload = download ?: -1L
        profile.subscriptionTotal = total ?: -1L
        profile.subscriptionExpireAt = expireAtMillis ?: 0L
        profile.subscriptionUpdateIntervalMinutes = updateIntervalMinutes ?: 0
        profile.subscriptionWebPageURL = webPageUrl.orEmpty()
    }
}

object SubscriptionMetadataParser {
    private const val MIN_UPDATE_INTERVAL_MINUTES = 15
    private const val SECONDS_TO_MILLIS = 1000L
    private const val MAX_COMMENT_LINES = 64

    fun parse(headers: Map<String, String>, body: String? = null): SubscriptionMetadata? {
        val valuesByHeader = parseCommentMetadata(body).toMutableMap()
        headers.forEach { (key, value) ->
            if (key.isMetadataKey() && value.isNotBlank()) {
                valuesByHeader[key.lowercase()] = value.trim()
            }
        }

        val userInfo =
            valuesByHeader["subscription-userinfo"]
        val updateInterval =
            valuesByHeader["profile-update-interval"]
        val webPageUrl =
            valuesByHeader["profile-web-page-url"]

        val values =
            userInfo.orEmpty()
                .split(';')
                .mapNotNull { item ->
                    val separator = item.indexOf('=')
                    if (separator <= 0) return@mapNotNull null
                    item.substring(0, separator).trim().lowercase() to
                        item.substring(separator + 1).trim()
                }.toMap()

        val metadata =
            SubscriptionMetadata(
                upload = values["upload"].toNonNegativeLong(),
                download = values["download"].toNonNegativeLong(),
                total = values["total"].toNonNegativeLong(),
                expireAtMillis = values["expire"].toEpochMillis(),
                updateIntervalMinutes = updateInterval.toIntervalMinutes(),
                webPageUrl = webPageUrl.toSafeWebPageUrl(),
            )
        return metadata.takeUnless { it.isEmpty() }
    }

    private fun parseCommentMetadata(body: String?): Map<String, String> {
        if (body.isNullOrBlank()) return emptyMap()
        return body.lineSequence()
            .take(MAX_COMMENT_LINES)
            .mapNotNull { line ->
                val comment = line.trimStart().takeIf { it.startsWith('#') }
                    ?.removePrefix("#")
                    ?.trim()
                    ?: return@mapNotNull null
                val separator = comment.indexOf(':')
                if (separator <= 0) return@mapNotNull null
                val key = comment.substring(0, separator).trim().lowercase()
                val value = comment.substring(separator + 1).trim()
                if (!key.isMetadataKey() || value.isEmpty()) return@mapNotNull null
                key to value
            }.toMap()
    }

    private fun String.isMetadataKey(): Boolean = this.equals("subscription-userinfo", ignoreCase = true) ||
        this.equals("profile-update-interval", ignoreCase = true) ||
        this.equals("profile-web-page-url", ignoreCase = true)

    private fun String?.toNonNegativeLong(): Long? = this?.toLongOrNull()?.takeIf { it >= 0 }

    private fun String?.toEpochMillis(): Long? {
        val seconds = this?.toLongOrNull()?.takeIf { it > 0 } ?: return null
        if (seconds > Long.MAX_VALUE / SECONDS_TO_MILLIS) return null
        return seconds * SECONDS_TO_MILLIS
    }

    private fun String?.toIntervalMinutes(): Int? {
        val hours = this?.trim()?.toLongOrNull()?.takeIf { it > 0 } ?: return null
        val minutes =
            if (hours > Int.MAX_VALUE.toLong() / 60L) {
                Int.MAX_VALUE.toLong()
            } else {
                max(MIN_UPDATE_INTERVAL_MINUTES.toLong(), hours * 60L)
            }
        return minutes.coerceAtMost(Int.MAX_VALUE.toLong()).toInt()
    }

    private fun String?.toSafeWebPageUrl(): String? {
        val value = this?.trim().orEmpty()
        if (value.isEmpty()) return null
        val uri = runCatching { URI(value) }.getOrNull() ?: return null
        if (uri.scheme.equals("http", ignoreCase = true).not() &&
            uri.scheme.equals("https", ignoreCase = true).not()
        ) {
            return null
        }
        if (uri.host.isNullOrBlank() || uri.rawUserInfo != null) return null
        return value
    }
}

File: iOS/Library/Network/SubscriptionMetadata.swift
import Foundation

public struct SubscriptionMetadata: Equatable, Sendable {
    public let upload: Int64?
    public let download: Int64?
    public let total: Int64?
    public let expireAt: Date?
    public let updateIntervalMinutes: Int32?
    public let webPageURL: String?

    public init(
        upload: Int64? = nil,
        download: Int64? = nil,
        total: Int64? = nil,
        expireAt: Date? = nil,
        updateIntervalMinutes: Int32? = nil,
        webPageURL: String? = nil
    ) {
        self.upload = upload
        self.download = download
        self.total = total
        self.expireAt = expireAt
        self.updateIntervalMinutes = updateIntervalMinutes
        self.webPageURL = webPageURL
    }

    public var isEmpty: Bool {
        upload == nil &&
            download == nil &&
            total == nil &&
            expireAt == nil &&
            updateIntervalMinutes == nil &&
            webPageURL == nil
    }
}

public enum SubscriptionMetadataParser {
    private static let supportedKeys = [
        "subscription-userinfo",
        "profile-update-interval",
        "profile-web-page-url",
    ]
    private static let maximumCommentLines = 64

    public static func parse(headers: [String: String], body: String? = nil) -> SubscriptionMetadata? {
        var metadataValues = parseCommentMetadata(body)
        for (key, value) in headers {
            let normalizedKey = key.lowercased()
            let normalizedValue = value.trimmingCharacters(in: .whitespacesAndNewlines)
            if supportedKeys.contains(normalizedKey), !normalizedValue.isEmpty {
                metadataValues[normalizedKey] = normalizedValue
            }
        }

        var userInfoValues = [String: String]()
        for item in metadataValues["subscription-userinfo", default: ""].split(separator: ";") {
            guard let separator = item.firstIndex(of: "=") else {
                continue
            }
            let key = item[..<separator].trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            let value = item[item.index(after: separator)...].trimmingCharacters(in: .whitespacesAndNewlines)
            if !key.isEmpty {
                userInfoValues[key] = value
            }
        }

        let metadata = SubscriptionMetadata(
            upload: nonnegativeInt64(userInfoValues["upload"]),
            download: nonnegativeInt64(userInfoValues["download"]),
            total: nonnegativeInt64(userInfoValues["total"]),
            expireAt: expirationDate(userInfoValues["expire"]),
            updateIntervalMinutes: updateIntervalMinutes(metadataValues["profile-update-interval"]),
            webPageURL: safeWebPageURL(metadataValues["profile-web-page-url"])
        )
        return metadata.isEmpty ? nil : metadata
    }

    private static func parseCommentMetadata(_ body: String?) -> [String: String] {
        guard let body, !body.isEmpty else {
            return [:]
        }
        var result = [String: String]()
        for line in body.split(omittingEmptySubsequences: false, whereSeparator: \.isNewline).prefix(maximumCommentLines) {
            let trimmedLine = line.trimmingCharacters(in: .whitespaces)
            guard trimmedLine.hasPrefix("#") else {
                continue
            }
            let comment = trimmedLine.dropFirst().trimmingCharacters(in: .whitespaces)
            guard let separator = comment.firstIndex(of: ":") else {
                continue
            }
            let key = comment[..<separator].trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            let value = comment[comment.index(after: separator)...].trimmingCharacters(in: .whitespacesAndNewlines)
            if supportedKeys.contains(key), !value.isEmpty {
                result[key] = value
            }
        }
        return result
    }

    private static func nonnegativeInt64(_ value: String?) -> Int64? {
        guard let value, let parsed = Int64(value), parsed >= 0 else {
            return nil
        }
        return parsed
    }

    private static func expirationDate(_ value: String?) -> Date? {
        guard let seconds = nonnegativeInt64(value), seconds > 0 else {
            return nil
        }
        let timestamp = TimeInterval(seconds)
        guard timestamp.isFinite else {
            return nil
        }
        return Date(timeIntervalSince1970: timestamp)
    }

    private static func updateIntervalMinutes(_ value: String?) -> Int32? {
        guard let value, let hours = Int64(value.trimmingCharacters(in: .whitespacesAndNewlines)), hours > 0 else {
            return nil
        }
        let maximumHours = Int64(Int32.max) / 60
        if hours > maximumHours {
            return Int32.max
        }
        return Int32(max(15, hours * 60))
    }

    private static func safeWebPageURL(_ value: String?) -> String? {
        guard let value else {
            return nil
        }
        let normalizedValue = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard
            let components = URLComponents(string: normalizedValue),
            let scheme = components.scheme?.lowercased(),
            scheme == "http" || scheme == "https",
            components.host?.isEmpty == false,
            components.user == nil,
            components.password == nil
        else {
            return nil
        }
        return normalizedValue
    }
}
